source/plugins/go/input/lib/cadvisor.go

```diff
@@ -0,0 +1,1219 @@
+
+func getSummaryStatsFromCAdvisor(winNode map[string]string) (*http.Response, error) {
+	relativeUri := "/stats/summary"
+	return getResponse(winNode, relativeUri)
+}
+
+// func getCongifzCAdvisor(winNode map[string]string) (*http.Response, error) {
+// 	relativeUri := "/configz"
+// 	return getResponse(winNode, relativeUri)
+// }
+
+// func getAllMetricsCAdvisor(winNode map[string]string) (*http.Response, error) {
+// 	relativeUri := "/metrics/cadvisor"
+// 	return getResponse(winNode, relativeUri)
+// }
+
+func GetPodsFromCAdvisor(winNode map[string]string) (*http.Response, error) {
+	relativeUri := "/pods"
+	return getResponse(winNode, relativeUri)
+}
+
+func getBaseCAdvisorUrl(winNode map[string]string) string {
+	cAdvisorSecurePort := isCAdvisorOnSecurePort()
+
+	var defaultHost string
+	if cAdvisorSecurePort {
+		defaultHost = fmt.Sprintf("https://localhost:%s", CADVISOR_SECURE_PORT)
+	} else {
+		defaultHost = fmt.Sprintf("http://localhost:%s", CADVISOR_NON_SECURE_PORT)
+	}
+
+	var nodeIP string
+	if winNode != nil {
+		nodeIP = winNode["InternalIP"]
+	} else {
+		nodeIP = os.Getenv("NODE_IP")
+	}
+
+	if nodeIP != "" {
+		Log.Infof("Using %s for CAdvisor Host", nodeIP)
+		if cAdvisorSecurePort {
+			return fmt.Sprintf("https://%s:%s", nodeIP, CADVISOR_SECURE_PORT)
+		} else {
+			return fmt.Sprintf("http://%s:%s", nodeIP, CADVISOR_NON_SECURE_PORT)
+		}
+	} else {
+		Log.Warnf("NODE_IP environment variable not set. Using default as: %s", defaultHost)
+		if winNode != nil {
+			return ""
+		} else {
+			return defaultHost
+		}
+	}
+}
+
+func getCAdvisorUri(winNode map[string]string, relativeUri string) string {
+	baseCAdvisorUrl := getBaseCAdvisorUrl(winNode)
+	if baseCAdvisorUrl == "" {
+		return ""
+	}
+	return fmt.Sprintf("%s%s", baseCAdvisorUrl, relativeUri)
+}
+
+func isCAdvisorOnSecurePort() bool {
+	cAdvisorSecurePort := false
+	if os.Getenv("IS_SECURE_CADVISOR_PORT") == "true" {
+		cAdvisorSecurePort = true
+	}
+	return cAdvisorSecurePort
+}
+
+func getResponse(winNode map[string]string, relativeUri string) (*http.Response, error) {
+	var response *http.Response
+	Log.Infof("Getting CAdvisor Uri Response")
+	bearerToken, _ := ioutil.ReadFile(BEARER_TOKEN_LOCATION)
+
+	cAdvisorUri := getCAdvisorUri(winNode, relativeUri)
+	Log.Infof("CAdvisor Uri: %s", cAdvisorUri)
+
+	if cAdvisorUri != "" {
+		uri, err := url.Parse(cAdvisorUri)
+		if err != nil {
+			Log.Errorf("Failed to parse CAdvisor Uri: %s", err)
+			return nil, err
+		}
+
+		var httpClient *http.Client
+		if isCAdvisorOnSecurePort() {
+			caCert, err := ioutil.ReadFile(CERT_LOCATION)
+			if err != nil {
+				Log.Errorf("Failed to read CA cert: %s", err)
+				return nil, err
+			}
+
+			caCertPool := x509.NewCertPool()
+			caCertPool.AppendCertsFromPEM(caCert)
+
+			httpClient = &http.Client{
+				Transport: &http.Transport{
+					Proxy: http.ProxyFromEnvironment,
+					DialContext: (&net.Dialer{
+						Timeout:   20 * time.Second,
+						KeepAlive: 0,
```

Comment:
any rationale to have KeepAlive: 0 here? should we consider to have KeepAlive for better performance since we will be making call every 1 min? wouldnt that help to avoid connection establishment every min?

Reply:
no particular reason. will remove